core/src/bank_forks.rs

```diff
@@ -27,6 +27,47 @@ impl BankForks {
             working_bank,
         }
     }
+
+    //produce a flat map of banks to all of its parents
+    pub fn flat_parents(&self) -> HashMap<u64, HashSet<u64>> {
+        let mut flat_parents = HashMap::new();
+        let mut pending: Vec<Arc<Bank>> = self.banks.values().cloned().collect();
+        while !pending.is_empty() {
+            let bank = pending.pop().unwrap();
+            if flat_parents.get(&bank.slot()).is_some() {
+                continue;
+            }
+            let set = bank.parents().into_iter().map(|b| b.slot()).collect();
+            flat_parents.insert(bank.slot(), set);
+            pending.extend(bank.parents().into_iter());
```

Comment:
At this point, you already know all the parents for each of the bank.parents() right? If we find the branch 1 -> 2 ->3 -> 4 for instance ,then we know that the parents of N are all M < N, so we could directly insert those into the map

Reply:
@carllin yep, that would work.  I just wanted to keep it simple for now.

Reply:
i think we can optimize this in a later pr